### ISSUE TITLE
MuninnPageCursor.assertPagedFileStillMappedAndGetIdOfLastPage throws NullPointerException and db goes into bugged state

### ISSUE THREAD

**Issue by rickardoberg** [neo4j/neo4j#11999]
- Neo4j version: 3.4.3 and 3.4.5
- Operating system: Ubuntu 16.04
- API/Driver: Cypher with embedded API

We randomly get a NPE in MuninnPageCursor.assertPagedFileStillMappedAndGetIdOfLastPage, can't see any pattern in which specific query triggers it. Once this happens other queries start failing randomly, i.e. it looks like the database doesn't recover from this. A restart fixes it, but the issue comes back usually within a day for us.

Stack trace:
```
Caused by: java.lang.NullPointerException: null
	at org.neo4j.io.pagecache.impl.muninn.MuninnPageCursor.assertPagedFileStillMappedAndGetIdOfLastPage(MuninnPageCursor.java:432)
	at org.neo4j.io.pagecache.impl.muninn.MuninnReadPageCursor.next(MuninnReadPageCursor.java:56)
	at org.neo4j.io.pagecache.impl.muninn.MuninnPageCursor.next(MuninnPageCursor.java:140)
	at org.neo4j.kernel.impl.store.CommonAbstractStore.readIntoRecord(CommonAbstractStore.java:1052)
	at org.neo4j.kernel.impl.store.CommonAbstractStore.getRecordByCursor(CommonAbstractStore.java:1035)
	at org.neo4j.kernel.impl.store.RelationshipGroupStore.getRecordByCursor(RelationshipGroupStore.java:35)
	at org.neo4j.kernel.impl.newapi.AllStoreHolder.group(AllStoreHolder.java:741)
	at org.neo4j.kernel.impl.newapi.DefaultRelationshipGroupCursor.next(DefaultRelationshipGroupCursor.java:169)
	at org.neo4j.internal.kernel.api.helpers.RelationshipDenseSelection.loopOnRelationship(RelationshipDenseSelection.java:207)
	at org.neo4j.internal.kernel.api.helpers.RelationshipDenseSelection.fetchNext(RelationshipDenseSelection.java:189)
	at org.neo4j.internal.kernel.api.helpers.RelationshipDenseSelectionCursor.next(RelationshipDenseSelectionCursor.java:30)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext$$anon$6.fetchNext(TransactionBoundQueryContext.scala:192)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext$$anon$6.fetchNext(TransactionBoundQueryContext.scala:189)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext$CursorIterator.<init>(TransactionBoundQueryContext.scala:1002)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext$$anon$6.<init>(TransactionBoundQueryContext.scala:189)
	at org.neo4j.cypher.internal.runtime.interpreted.TransactionBoundQueryContext.getRelationshipsForIds(TransactionBoundQueryContext.scala:189)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext$$anonfun$getRelationshipsForIds$1.apply(ExceptionTranslatingQueryContext.scala:220)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext$$anonfun$getRelationshipsForIds$1.apply(ExceptionTranslatingQueryContext.scala:220)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslationSupport$class.translateException(ExceptionTranslationSupport.scala:33)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext.translateException(ExceptionTranslatingQueryContext.scala:41)
	at org.neo4j.cypher.internal.compatibility.v3_4.ExceptionTranslatingQueryContext.getRelationshipsForIds(ExceptionTranslatingQueryContext.scala:220)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.ExpandAllPipe$$anonfun$internalCreateResults$1.apply(ExpandAllPipe.scala:43)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.ExpandAllPipe$$anonfun$internalCreateResults$1.apply(ExpandAllPipe.scala:40)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:435)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:441)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:104)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:48)
	at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:310)
	at scala.collection.AbstractIterator.to(Iterator.scala:1334)
	at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:302)
	at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1334)
	at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:289)
	at scala.collection.AbstractIterator.toArray(Iterator.scala:1334)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.RollUpApplyPipe$$anonfun$internalCreateResults$1.apply(RollUpApplyPipe.scala:40)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.RollUpApplyPipe$$anonfun$internalCreateResults$1.apply(RollUpApplyPipe.scala:33)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1334)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.EagerAggregationPipe.internalCreateResults(EagerAggregationPipe.scala:118)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:76)
	at org.neo4j.cypher.internal.runtime.interpreted.pipes.PipeWithSource.createResults(Pipe.scala:72)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.executionplan.BaseExecutionResultBuilderFactory$BaseExecutionWorkflowBuilder.createResults(DefaultExecutionResultBuilderFactory.scala:105)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.executionplan.BaseExecutionResultBuilderFactory$BaseExecutionWorkflowBuilder.build(DefaultExecutionResultBuilderFactory.scala:77)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.BuildInterpretedExecutionPlan$$anonfun$getExecutionPlanFunction$1.apply(BuildInterpretedExecutionPlan.scala:97)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.BuildInterpretedExecutionPlan$$anonfun$getExecutionPlanFunction$1.apply(BuildInterpretedExecutionPlan.scala:80)
	at org.neo4j.cypher.internal.compatibility.v3_4.runtime.BuildInterpretedExecutionPlan$InterpretedExecutionPlan.run(BuildInterpretedExecutionPlan.scala:111)
	at org.neo4j.cypher.internal.compatibility.LatestRuntimeVariablePlannerCompatibility$ExecutionPlanWrapper$$anonfun$run$1.apply(LatestRuntimeVariablePlannerCompatibility.scala:128)
	at org.neo4j.cypher.internal.compatibility.LatestRuntimeVariablePlannerCompatibility$ExecutionPlanWrapper$$anonfun$run$1.apply(LatestRuntimeVariablePlannerCompatibility.scala:124)
	at org.neo4j.cypher.exceptionHandler$runSafely$.apply(exceptionHandler.scala:89)
	at org.neo4j.cypher.internal.compatibility.LatestRuntimeVariablePlannerCompatibility$ExecutionPlanWrapper.run(LatestRuntimeVariablePlannerCompatibility.scala:124)
	at org.neo4j.cypher.internal.PreparedPlanExecution.execute(PreparedPlanExecution.scala:29)
	at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:119)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:61)
	at org.neo4j.kernel.impl.factory.ClassicCoreSPI.executeQuery(ClassicCoreSPI.java:79)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacade.execute(GraphDatabaseFacade.java:428)
	at org.neo4j.kernel.impl.factory.GraphDatabaseFacade.execute(GraphDatabaseFacade.java:411)
```

Looking at commits to the MuninnPageCursor, it feels like the assert is done wrongly, as on close the reference becomes null, but this is not checked for. Here's a commit that seems relevant:
https://github.com/neo4j/neo4j/commit/c86c236cb6a614e6cf024340b64b6ad181f22a7b

Workaround: until this is fixed our workaround is to restart all servers daily.

**Comment by chrisvest** [comment]
Are you doing funny things with threading, like getting a Cypher Result in one thread and consuming or closing it in another? We are suspecting the cursor caching. Nulling out that field is actually a safe guard against using closed cursors.

**Comment by rickardoberg** [comment]
Hi Chris!

Yes, I am doing exactly that. I have a wrapper on top of
GraphDatabaseService that runs queries in a thread pool and makes result a
CompletableFuture, so that we can easily do parallel queries, or do queries
that requires reading new state. Example: first query that returns lots of
rows, issue writes in a separate thread (we have single writer thread),
then be able to read those changes while still having the first query open.
Another example is issuing two parallel queries and join results, to avoid
having to do serial queries even when they are independent.

This used to work just fine in 3.3. Is it possible to fix in 3.4, or do I
have to rearchitect our Neo4j usage?

cheers, Rickard

**Comment by rickardoberg** [comment]
I'm going to check straight away if we can refactor away the cases where we need threading, and make same-thread usage the default instead.

**Comment by chrisvest** [comment]
The probabilities changed in 3.4 but the problem is also there in 3.3 and older. The REST server runs into it on occasion, though it’s quite rare. It’s being worked on.

**Comment by rickardoberg** [comment]
Ok, thanks. Then for now I will focus on changing our wrapper to be same-thread, and handle the cases where we need threading explicitly in client code, so that it has same-thread handling when it comes to Neo4j. Thanks!

**Comment by chrisvest** [comment]
This should be fixed in 3.4.8 by "Remove cursor pooling on a page cache level" https://github.com/neo4j/neo4j/wiki/Neo4j-3.4-changelog#348
